### ISSUE TITLE
Swarm images not pulling automatically using registry auth

### ISSUE THREAD

**Issue by mackeyja92** [docker/docker#29266]
**Description**
When updating a service that was created with `--with-registry-auth` and then updating the image with a `service update ... --with-registry-auth service_name` the other nodes fail to pull the new image.

**Steps to reproduce the issue:**
1. Create a new service with the `--with-registry-auth` tag
2. Create a new image on one of the machines
3. Push that new image up to your private repo
4. On that same machine issue the following 
```
docker service update --image $(docker inspect -f '{{index .RepoDigests 0}}' <link_to_private_repo_here>) --with-registry-auth <service_name>
```
**Describe the results you received:**
The other nodes were unable to pull the image. Looking at the service with `docker service ps <service_name>` gave the following

```
\_<service_name.3>  <link_to_private_repo>@sha256:<RepoDigest> <node_name>  Shutdown       Rejected about a minute ago  "No such image: <link_to_private_repo>
```
**Describe the results you expected:**
That the image on those nodes would pull automatically and update correctly.

**Additional information you deem important (e.g. issue happens only occasionally):**
If I don't touch the service but go to that node and just pull the image manually (without the repo digest) the service will continue to work after the next time docker swarm tries to start it again.

The only thing I can think of that may be off is that the `docker-service update...` command is being ran by my CI under the CI user on the machine. I believe the token that it is using might be the one provided as a one time token. However, the root user and the docker user are both logged into the registry. This might be the cause of the issue and if so the error message provided by docker could be clearer.

**Output of `docker version`:**

```
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:44:32 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:44:32 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 11
 Running: 3
 Paused: 0
 Stopped: 8
Images: 360
Server Version: 1.12.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 425
 Dirperm1 Supported: true
Logging Driver: syslog
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge overlay host
Swarm: active
 NodeID: 777xx34g1z9us78qc3fdw1jfh
 Is Manager: true
 ClusterID: 7i0dq8spowatrive3pisdv5mt
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.10.10.131
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 4.4.0-36-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 125.8 GiB
Name: docker1
ID: VCCQ:YEF4:QDY5:N44K:CFVL:2XKR:XEQJ:JEDY:FIO7:C3QW:AGOX:GVXD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Comment by cpuguy83** [comment]
ping @aaronlehmann

**Comment by aaronlehmann** [comment]
`--with-registry-auth` tells the "service update" command to update registry credentials. If you don't want to touch the credentials, it's better to omit this option. Can you try without `--with-registry-auth` to see if this corrects the problem?

**Comment by mackeyja92** [comment]
@aaronlehmann I just removed the `--with-registry-auth` from the update command and still experiencing the same issue.

**Comment by aaronlehmann** [comment]
> The only thing I can think of that may be off is that the `docker-service update...` command is being ran by my CI under the CI user on the machine. I believe the token that it is using might be the one provided as a one time token. However, the root user and the docker user are both logged into the registry. This might be the cause of the issue and if so the error message provided by docker could be clearer.

It would be great to try and verify that this is the issue. Perhaps you could try the `docker service update --with-registry-auth ...` command from a user account that is logged in without using a one-time token and check if the worker nodes are able to pull after that.

If this is indeed the problem, I think Docker 1.13 will provide a better error message in this case in most cases. On a "service update", it will try to contact the registry to resolve the image to a digest. If this fails, an warning message gets printed (but the service update still goes ahead). In your particular case, you're already specifying a digest, so this step would be skipped. Perhaps we should try to contact the registry anyway just so we can warn if the credentials don't work.

**Comment by mackeyja92** [comment]
I manually built the image, pushed it and then then did the update command as the root user on the machine which is logged it without `--with-registry-auth` and it did not work. I went and pulled the image on the other machines and did the exact same update from the first machine again and they started right up.

Then, I did the same thing but added in the `--with-registry-auth` and it worked fabulously. So it is just not showing the correct error message. 

Also, why is it that it wouldn't use the registry auth used to create the service? That one would be the same one I just used to update it. Does it only use those auth when it initially tries to pull the image? Why can't it use that auth when updating?
